xarray/backends/netCDF4_.py

```diff
@@ -25,35 +25,33 @@
                   '|': 'native'}
 
 
-class BaseNetCDF4Array(NDArrayMixin):
+class BaseNetCDF4Array(NdimSizeLenMixin):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
+        # don't autoclose yet, because the file is already open for reading
+        # variable metadata
+        array = self.get_array(autoclose=False)
+        self.shape = array.shape
```

Comment:
This made me a little nervous about our test coverage for the `dtype` property on unloaded arrays, see #1311.